midend/simplifyBitwise.cpp

```diff
@@ -0,0 +1,98 @@
+#include "simplifyBitwise.h"
+
+namespace P4 {
+
+bool SimplifyBitwise::Scan::preorder(const IR::AssignmentStatement *as) {
+    if (!as->right->is<IR::BOr>())
+        return false;
+    can_be_changed = true;
+    mpz_init(total_mask.get_mpz_t());
+    return true;
+}
+
+bool SimplifyBitwise::Scan::preorder(const IR::BOr *bor) {
+    if (!findContext<IR::AssignmentStatement>())
+        can_be_changed = false;
+    if (!(bor->left->is<IR::BAnd>() && bor->right->is<IR::BAnd>()))
+        can_be_changed = false;
+    return can_be_changed;
+}
+
+bool SimplifyBitwise::Scan::preorder(const IR::BAnd *band) {
+    if (!findContext<IR::BOr>())
+        can_be_changed = false;
+    if (!band->left->is<IR::Constant>() && !band->right->is<IR::Constant>())
+        can_be_changed = false;
+    if (band->left->is<IR::Constant>() && band->right->is<IR::Constant>())
+        can_be_changed = false;
+    return can_be_changed;
+}
+
+bool SimplifyBitwise::Scan::preorder(const IR::Constant *constant) {
+    if (!findContext<IR::BAnd>()) {
+        can_be_changed = false;
+        return false;
+    }
+    // Ensure that there are no collisions within the bitmask
+    mpz_class mask;
+    mpz_init(mask.get_mpz_t());
+    mpz_class zero;
+    mpz_init(zero.get_mpz_t());
+    mpz_and(mask.get_mpz_t(), total_mask.get_mpz_t(), constant->value.get_mpz_t());
+    if (mpz_cmp(mask.get_mpz_t(), zero.get_mpz_t()) == 0) {
+        mpz_ior(total_mask.get_mpz_t(), mask.get_mpz_t(), total_mask.get_mpz_t());
```

Comment:
if (mask == 0) {
        total_mask = mask | total_mask;

...which is a noop?

Reply:
You're right, mask should be replaced by constant->value.